File: app/utils/dawproject_python/clip.py
from lxml import etree as ET
from classes.nameable import Nameable
from classes.referenceable import Referenceable
from classes.timeline import Timeline
from classes.timeUnit import TimeUnit


class Clip(Nameable):
    def __init__(
        self,
        time,
        duration=None,
        content_time_unit=None,
        play_start=None,
        play_stop=None,
        loop_start=None,
        loop_end=None,
        fade_time_unit=None,
        fade_in_time=None,
        fade_out_time=None,
        content=None,
        reference=None,
        name=None,
        color=None,
        comment=None,
    ):
        super().__init__(name, color, comment)
        self.time = time
        self.duration = duration
        self.content_time_unit = content_time_unit
        self.play_start = play_start
        self.play_stop = play_stop
        self.loop_start = loop_start
        self.loop_end = loop_end
        self.fade_time_unit = fade_time_unit
        self.fade_in_time = fade_in_time
        self.fade_out_time = fade_out_time
        self.content = content
        self.reference = reference

    def to_xml(self):
        # Create the Clip XML element
        clip_elem = ET.Element("Clip")

        # Set common attributes
        clip_elem.set("time", str(self.time))
        if self.duration is not None:
            clip_elem.set("duration", str(self.duration))
        if self.content_time_unit is not None:
            clip_elem.set("contentTimeUnit", self.content_time_unit.value)
        if self.play_start is not None:
            clip_elem.set("playStart", str(self.play_start))
        if self.play_stop is not None:
            clip_elem.set("playStop", str(self.play_stop))
        if self.loop_start is not None:
            clip_elem.set("loopStart", str(self.loop_start))
        if self.loop_end is not None:
            clip_elem.set("loopEnd", str(self.loop_end))
        if self.fade_time_unit is not None:
            clip_elem.set("fadeTimeUnit", self.fade_time_unit.value)
        if self.fade_in_time is not None:
            clip_elem.set("fadeInTime", str(self.fade_in_time))
        if self.fade_out_time is not None:
            clip_elem.set("fadeOutTime", str(self.fade_out_time))

        # Append content if present
        if self.content is not None:
            content_elem = self.content.to_xml()
            clip_elem.append(content_elem)

        # Reference handling
        if self.reference is not None:
            clip_elem.set("reference", self.reference.id)

        # Append inherited attributes
        nameable_elem = super().to_xml()
        for key, value in nameable_elem.attrib.items():
            clip_elem.set(key, value)

        return clip_elem

    @classmethod
    def from_xml(cls, element):
        # Initialize using parent class method
        instance = super().from_xml(element)

        # Extract Clip-specific attributes
        instance.time = float(element.get("time"))
        instance.duration = (
            float(element.get("duration")) if element.get("duration") else None
        )
        instance.content_time_unit = (
            TimeUnit(element.get("contentTimeUnit"))
            if element.get("contentTimeUnit")
            else None
        )
        instance.play_start = (
            float(element.get("playStart")) if element.get("playStart") else None
        )
        instance.play_stop = (
            float(element.get("playStop")) if element.get("playStop") else None
        )
        instance.loop_start = (
            float(element.get("loopStart")) if element.get("loopStart") else None
        )
        instance.loop_end = (
            float(element.get("loopEnd")) if element.get("loopEnd") else None
        )
        instance.fade_time_unit = (
            TimeUnit(element.get("fadeTimeUnit"))
            if element.get("fadeTimeUnit")
            else None
        )
        instance.fade_in_time = (
            float(element.get("fadeInTime")) if element.get("fadeInTime") else None
        )
        instance.fade_out_time = (
            float(element.get("fadeOutTime")) if element.get("fadeOutTime") else None
        )

        # Handling content and reference
        content_elem = element.find("Timeline")
        if content_elem is not None:
            content_class = globals().get(content_elem.tag)
            if content_class and issubclass(content_class, Timeline):
                instance.content = content_class.from_xml(content_elem)

        reference_id = element.get("reference")
        if reference_id:
            instance.reference = Referenceable.get_by_id(reference_id)

        return instance

File: app/utils/dawproject-new/clip.ts
// dawproject/clip.ts
import { Audio } from "./audio";
import { ClipSlot } from "./clip-slot";
import { Clips } from "./clips";
import { Lanes } from "./lanes";
import type { Markers } from "./markers";
import { Nameable } from "./nameable";
import { Notes } from "./notes";
import type { Points } from "./points";
import type {
  Clip as ClipType,
  TimeUnit,
  XsDouble,
  XsString,
} from "./project-schema";
import { Timeline } from "./timeline";
import { Video } from "./video";
import { Warps } from "./warps";

// Define union type for possible child elements as per XSD choice
export type ClipContent =
  | Timeline
  | Lanes
  | Notes
  | Clips
  | ClipSlot
  | Markers
  | Warps
  | Audio
  | Video
  | Points;

/**
 * Represents a clip on a timeline lane.
 * Corresponds to the 'clip' complex type in Project.xsd.
 * Inherits attributes and child elements from Nameable.
 */
export class Clip extends Nameable implements ClipType {
  // XML attributes are prefixed with '@_'

  /**
   * The start time of the clip relative to its parent timeline.
   * (Required attribute - xs:double)
   */
  public "@_time": XsDouble;

  /**
   * The duration of the clip.
   * (Optional attribute - xs:double)
   */
  public "@_duration"?: XsDouble;

  /**
   * The time unit used for the clip's content (e.g., beats, seconds).
   * (Optional attribute - timeUnit enum)
   */
  public "@_contentTimeUnit"?: TimeUnit;

  /**
   * The start time for playback within the clip's content.
   * (Optional attribute - xs:double)
   */
  public "@_playStart"?: XsDouble;

  /**
   * The stop time for playback within the clip's content.
   * (Optional attribute - xs:double)
   */
  public "@_playStop"?: XsDouble;

  /**
   * The start time for looping within the clip's content.
   * (Optional attribute - xs:double)
   */
  public "@_loopStart"?: XsDouble;

  /**
   * The end time for looping within the clip's content.
   * (Optional attribute - xs:double)
   */
  public "@_loopEnd"?: XsDouble;

  /**
   * The time unit used for fade times.
   * (Optional attribute - timeUnit enum)
   */
  public "@_fadeTimeUnit"?: TimeUnit;

  /**
   * The duration of the fade-in.
   * (Optional attribute - xs:double)
   */
  public "@_fadeInTime"?: XsDouble;

  /**
   * The duration of the fade-out.
   * (Optional attribute - xs:double)
   */
  public "@_fadeOutTime"?: XsDouble;

  /**
   * A reference to another element (e.g., a shared clip definition).
   * (Optional attribute - xs:IDREF)
   */
  public "@_reference"?: XsString;

  // Property corresponding to child element choice

  /**
   * The content of the clip (e.g., notes, audio, automation).
   * (Optional child element - choice)
   */
  public Timeline?: Timeline;
  public Lanes?: Lanes;
  public Notes?: Notes;
  public Clips?: Clips;
  public ClipSlot?: ClipSlot;
  public Markers?: Markers;
  public Warps?: Warps;
  public Audio?: Audio;
  public Video?: Video;
  public Points?: Points;

  /**
   * @param time - The start time of the clip relative to its parent timeline. (Required attribute - xs:double)
   * @param duration - The duration of the clip. (Optional attribute - xs:double)
   * @param name - The name of the clip. (Optional attribute inherited from Nameable)
   * @param color - The color of the clip. (Optional attribute inherited from Nameable)
   * @param comment - A comment for the clip. (Optional attribute inherited from Nameable)
   * @param contentTimeUnit - The time unit used for the clip's content. (Optional attribute - timeUnit enum)
   * @param playStart - The start time for playback within the clip's content. (Optional attribute - xs:double)
   * @param playStop - The stop time for playback within the clip's content. (Optional attribute - xs:double)
   * @param loopStart - The start time for looping within the clip's content. (Optional attribute - xs:double)
   * @param loopEnd - The end time for looping within the clip's content. (Optional attribute - xs:double)
   * @param fadeTimeUnit - The time unit used for fade times. (Optional attribute - timeUnit enum)
   * @param fadeInTime - The duration of the fade-in. (Optional attribute - xs:double)
   * @param fadeOutTime - The duration of the fade-out. (Optional attribute - xs:double)
   * @param reference - A reference to another element. (Optional attribute - xs:IDREF)
   */
  constructor(
    time: XsDouble,
    duration?: XsDouble,
    name?: XsString,
    color?: XsString,
    comment?: XsString,
    contentTimeUnit?: TimeUnit,
    playStart?: XsDouble,
    playStop?: XsDouble,
    loopStart?: XsDouble,
    loopEnd?: XsDouble,
    fadeTimeUnit?: TimeUnit,
    fadeInTime?: XsDouble,
    fadeOutTime?: XsDouble,
    reference?: XsString
  ) {
    super(name, color, comment);
    this["@_time"] = time;
    this["@_duration"] = duration;
    this["@_contentTimeUnit"] = contentTimeUnit;
    this["@_playStart"] = playStart;
    this["@_playStop"] = playStop;
    this["@_loopStart"] = loopStart;
    this["@_loopEnd"] = loopEnd;
    this["@_fadeTimeUnit"] = fadeTimeUnit;
    this["@_fadeInTime"] = fadeInTime;
    this["@_fadeOutTime"] = fadeOutTime;
    this["@_reference"] = reference;
  }

  /**
   * Sets the clip's content, ensuring only one content type is active.
   * This matches the XSD choice restriction where only one child element can be present.
   */
  public setContent(content: ClipContent): void {
    // Clear existing content
    this.Timeline = undefined;
    this.Lanes = undefined;
    this.Notes = undefined;
    this.Clips = undefined;
    this.ClipSlot = undefined;
    this.Markers = undefined;
    this.Warps = undefined;
    this.Audio = undefined;
    this.Video = undefined;
    this.Points = undefined;

    // Helper function to safely cast content
    const isInstance = <T>(obj: any, className: string): obj is T => {
      return obj && obj.constructor && obj.constructor.name === className;
    };

    // Set the new content
    if (isInstance<Timeline>(content, "Timeline")) {
      this.Timeline = content;
    } else if (isInstance<Lanes>(content, "Lanes")) {
      this.Lanes = content;
    } else if (isInstance<Notes>(content, "Notes")) {
      this.Notes = content;
    } else if (isInstance<Clips>(content, "Clips")) {
      this.Clips = content;
    } else if (isInstance<ClipSlot>(content, "ClipSlot")) {
      this.ClipSlot = content;
    } else if (isInstance<Markers>(content, "Markers")) {
      this.Markers = content;
    } else if (isInstance<Warps>(content, "Warps")) {
      this.Warps = content;
    } else if (isInstance<Audio>(content, "Audio")) {
      this.Audio = content;
    } else if (isInstance<Video>(content, "Video")) {
      this.Video = content;
    } else if (isInstance<Points>(content, "Points")) {
      this.Points = content;
    } else {
      throw new Error("Invalid content type for Clip");
    }
  }
}
